A68-random-subsetting.md

```diff
@@ -0,0 +1,117 @@
+A68: Random subsetting with rendezvous hashing LB policy.
+----
+* Author(s): @s-matyukevich
+* Approver: 
+* Status: Draft
+* Implemented in: Go, Java
+* Last updated: [Date]
+* Discussion at: 
+
+## Abstract
+
+Add support for the `random_subsetting` load balancing policy.
+
+## Background
+
+Currently, gRPC is lacking a way to select a subset of endpoints available from the resolver and load-balance requests between them. Out of the box, users have the choice between two extremes: `pick_first` which sends all requests to one random backend, and `round_robin` which sends requests to all available backends. `pick_first` has poor connection balancing when the number of client is not much higher than the number of servers. The problem is exacerbated during rollouts because `pick_first` does not change endpoint on resolver updates if the current subchannel remains `READY`. `round_robin` results in every server having as many connections open as there are clients, which is unnecessarily costly when there are many clients, and makes local load balancing decisions (such as outlier detection) less precise.
+
+### Related Proposals: 
+* [gRFC A27: A52: gRPC xDS Custom Load Balancer Configuration](https://github.com/grpc/proposal/blob/master/A52-xds-custom-lb-policies.md)
+
+## Proposal
+
+Introduce a new LB policy, `random_subsetting`. This policy selects a subset of addresses and passes them to the child LB policy. It maintains 2 important properties:
+* The policy tries to distribute connections among servers as equally as possible. The higher `(N_clients*subset_size)/N_servers` ratio is, the closer the resulting server connection distribution is to uniform.
+* The policy minimizes the amount of connection churn generated during server scale-ups by using [rendezvous hashing](https://en.wikipedia.org/wiki/Rendezvous_hashing)
+
+### Subsetting algorithm
+
+* The policy receives a single configuration parameter: `subset_size`, which must be configured by the user.
+* When the lb policy is initialized it also creates a random 32-byte long `salt` string. 
+* After every resolver update the policy picks a new subset. It does this by implementing `rendezvous hashing` algorithm:
+  * Concatenate string representation of every address in the list with `salt`.
+  * For every resulting entity compute [MurmurHash3](https://en.wikipedia.org/wiki/MurmurHash) hash, which produces 128-byte output.
+  * Sort all entities by hash.
+  * Pick first `subset_size` values from the list.
+* Pass the resulting subset to the child LB policy.
```

Comment:
Out of scope for this doc, but I'm curious:

How do LB policies compose? I suspect it depends on the grpc implementation, but do have clear abstraction boundaries separation between selection (e.g. heap, p2c, rr) and the load metric (e.g. orca, least loaded, etc.)  ?

Reply:
In Go a balancer is just an instance of the following interface https://github.com/grpc/grpc-go/blob/master/balancer/balancer.go#L371 If a balancer wants to redirect some work to a child balancer it should implement this logic itself. There are several ways how it can do this, for example:
* [outlierdetection](https://github.com/grpc/grpc-go/blob/master/xds/internal/balancer/outlierdetection/balancer.go#L155) LB maintains the list of sub channels. For every sub-channel it creates a wrapper so it can get notified whenever a error on a sub-channel occurs. Whenever a err rate on a particular sub-channel reaches some threshold it removes this sub-channel from the list and notifies child balancer that the address list got updated. From the point of view of child balancer this is identical to the situation when a resolver removes the same address because it is no longer available in DNS.
* What I am proposing in this gRFC is a lot simpler: we always redirect all work to the child balancer, but do some filtering whenever the address list got updated. And we don't maintain any state in the parent balancer. 
In both cases this logic lives in the balancer itself and grpc don't enforce how exactly communication between parent and child balancer should happen.